java/vector/src/main/java/org/apache/arrow/vector/dictionary/BatchedDictionary.java

```diff
@@ -0,0 +1,270 @@
+
+  private final boolean forFileIPC;
+
+  private int deltaIndex;
+
+  private int dictionaryIndex;
+
+  private boolean wasReset;
+
+  /**
+   * Creates a dictionary with two vectors of the given types. The dictionary vector
+   * will be named "{name}-dictionary".
+   * <p>
+   * To use this dictionary, provide the dictionary vector to a {@link DictionaryProvider},
+   * add the {@link #getIndexVector()} to the {@link org.apache.arrow.vector.VectorSchemaRoot}
+   * and call the {@link #setSafe(int, byte[], int, int)} or other set methods.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param forFileIPC Whether the data will be written to a file or stream IPC. Throws an
+   *                   exception if a replacement dictionary is provided to a file IPC.
+   * @param allocator The allocator to use.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      boolean forFileIPC,
+      BufferAllocator allocator
+  ) {
+    this(name, encoding, dictionaryType, indexType, forFileIPC, allocator, "-dictionary");
+  }
+
+  /**
+   * Creates a dictionary with two vectors of the given types.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param forFileIPC Whether the data will be written to a file or stream IPC. Throws an
+   *                   exception if a replacement dictionary is provided to a file IPC.
+   * @param allocator The allocator to use.
+   * @param suffix A non-null suffix to append to the name of the dictionary.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      boolean forFileIPC,
+      BufferAllocator allocator,
+      String suffix
+  ) {
+    this.encoding = encoding;
+    this.forFileIPC = forFileIPC;
+    FieldVector vector = new FieldType(false, dictionaryType, null)
+        .createNewSingleVector(name + suffix, allocator, null);
+    if (!(BaseVariableWidthVector.class.isAssignableFrom(vector.getClass()))) {
+      throw new IllegalArgumentException("Dictionary must be a superclass of 'BaseVariableWidthVector' " +
+          "such as 'VarCharVector'.");
+    }
+    dictionary = (BaseVariableWidthVector) vector;
+    vector = new FieldType(true, indexType, encoding)
+        .createNewSingleVector(name, allocator, null);
+    if (!(BaseIntVector.class.isAssignableFrom(vector.getClass()))) {
+      throw new IllegalArgumentException("Index vector must be a superclass type of 'BaseIntVector' " +
```

Comment:
It would be good to validate both the dictionaryType and indexType before creating either vectors so that the constructor never gets into a state where it throws with vectors already constructed (they should actually be released with close()).

Reply:
ah good catch.